src/distributed_join.cuh

```diff
@@ -0,0 +1,385 @@
+ *
+ * @param[in] buckets Received table from each rank during 'all_to_all_comm'. The device buffer
+ * inside will be freed by this function.
+ * @param[in] counts Number of items received from each rank, got from 'all_to_all_comm'.
+ * @param[in] dtypes Column types
+ *
+ * @return Table formed by merging all buckets in *buckets*.
+ */
+std::unique_ptr<table>
+all_to_all_merge_data(
+    vector<vector<void *> > &buckets,
+    vector<vector<int64_t> > &counts,
+    vector<cudf::data_type> &dtypes,
+    Communicator *communicator)
+{
+    vector<std::unique_ptr<column> > merged_table;
+
+    for (int icol = 0; icol < buckets.size(); icol++) {
+        size_t dtype_size = cudf::size_of(dtypes[icol]);
+
+        int64_t total_count;
+        rmm::device_buffer merged_data = merge_free_received_offset(
+            buckets[icol], counts[icol], dtype_size, total_count, communicator, false
+        );
+
+        merged_table.push_back(std::make_unique<column>(dtypes[icol], total_count, std::move(merged_data)));
+    }
+
+    return std::make_unique<table>(std::move(merged_table));
+}
+
+
+/**
+ * All-to-all communication with merging for a single batch.
+ *
+ * This function needs to be called collectively by all ranks in MPI_COMM_WORLD. For every ranks,
+ * all arguments are significant.
+ *
+ * @param[in] input Input table to be communicated.
+ * @param[in] offset Vector of length mpi_size + 1 such that offset[i] represents the starting row
+ * index of bucket i in `input`.
+ * @param[in] communicator An instance of `Communicator` used for communication.
+ *
+ * @return The communication result which gathers bucket i of `input` on rank i.
+ */
+std::unique_ptr<table>
+all_to_all_comm_single_batch(
+    cudf::table_view input,
+    vector<cudf::size_type> offset,
+    Communicator *communicator)
+{
+    vector<std::unique_ptr<column> > recv_columns;
+
+    for (cudf::size_type icol = 0; icol < input.num_columns(); icol++) {
+
+        std::size_t dtype_size = cudf::size_of(input.column(icol).type());
+
+        vector<comm_handle_t> send_requests = send_data_by_offset(
+            input.column(icol).head(), offset, dtype_size, communicator
+        );
+
+        vector<void *> recv_data;
+        vector<int64_t> count;
+
+        vector<comm_handle_t> recv_requests = recv_data_by_offset(
+            recv_data, count, dtype_size, communicator
+        );
+
+        communicator->waitall(send_requests);
+        communicator->waitall(recv_requests);
+
+        int64_t nrows;
+        rmm::device_buffer merged_data = merge_free_received_offset(
+            recv_data, count, dtype_size, nrows, communicator
+        );
+
+        recv_columns.push_back(std::make_unique<column>(
+            input.column(icol).type(), nrows, std::move(merged_data))
+        );
+    }
+
+    return std::make_unique<table>(std::move(recv_columns));
+}
+
+
+void
+inner_join_func(
+    vector<vector<vector<void *> > > &left_buckets,
```

Comment:
All of these non-const vectors looks fishy to me.

Reply:
Added `const` qualifiers to them in https://github.com/rapidsai/distributed-join/pull/10/commits/9ce2080062df4b8d58f2524be4c601f321328e1a